### PR TITLE
Furnace and recipe base

### PR DESCRIPTION
**Pull request by susch19**
* made a proof of concept working furnace with recipes
    * basic energy ↔ time conversion (Needs work)
    * use leafs, wood, and cotton as burnables
    * has 3 inventories, input, output, and one for production resources
* added a base of recipes to work with, currently most properties are not in use but planned for the future
* added meat for testing purposes
* implemented entity to entity interaction
* made inventory more configurable (max size, max slots / fixed slots, max volume)
* reworked extensions to support ui extensions and be more flexible all together
* added simple state machine for furnace

### CODE REVIEW

File: OctoAwesome/OctoAwesome.Client/Cache/VerticesForChunk.cs [hunk 1]
```diff
@@ -11,12 +11,20 @@ internal class VerticesForChunk : IDisposable, ISerializable
     {
         public int Version { get; set; }
         public Index3 ChunkPosition { get; set; }
-        public VertexPositionNormalTextureLight[]? Vertices { get; set; }
+        public VertexPositionNormalTextureLight[]? Vertices { get; set; }        
 
+        /// <summary>
+        /// Initializes a new instance of the<see cref="VerticesForChunk" /> class
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Cache/VerticesForChunk.cs]:
```suggestion
        /// Initializes a new instance of the <see cref="VerticesForChunk" /> class.
```

File: OctoAwesome/OctoAwesome.Client/Cache/VerticesForChunk.cs [hunk 2]
```diff
@@ -11,12 +11,20 @@ internal class VerticesForChunk : IDisposable, ISerializable
...
+        /// </summary>
         public VerticesForChunk()
         {
 
         }
+
+
+        /// <summary>
+        /// Initializes a new instance of the<see cref="VerticesForChunk" /> class
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Cache/VerticesForChunk.cs]:
```suggestion
        /// Initializes a new instance of the <see cref="VerticesForChunk" /> class.
```

File: OctoAwesome/OctoAwesome.Client/Program.cs [hunk 3]
```diff
@@ -20,6 +24,35 @@ public static class Program
...
+
+                        }
+                    }
+
+                }
+                else
+                    ;
+            }
+
+
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Program.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Client/Program.cs [hunk 4]
```diff
@@ -1,8 +1,12 @@
 ﻿#region Using Statements
 
 using OctoAwesome.Logging;
+
 using System;
+using System.Collections.Generic;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Program.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Client/Program.cs [hunk 5]
```diff
@@ -1,8 +1,12 @@
...
 using System.IO;
+using System.Text.Json;
+using System.Text.Json.Nodes;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Program.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome/StandaloneTypeContainer.cs [hunk 6]
```diff
@@ -124,11 +124,15 @@ public object GetUnregistered(Type type)
...
+                    else if (parameter.IsOptional && parameter.HasDefaultValue)
+                    {
+                        tmpList.Add(parameter.DefaultValue);
+                    }
+                    else // Unknown parameter type without default value
                     {
                         tmpList.Clear();
                         next = true;
-                        break;
+                        break;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/StandaloneTypeContainer.cs]:
```suggestion
                        break;
```

File: OctoAwesome/OctoAwesome/Simulation.cs [hunk 7]
```diff
@@ -48,50 +51,57 @@ public sealed class Simulation : IDisposable
...
+        //public IReadOnlyList<Entity> Entities => entities;
 
-        /// <summary>
-        /// Gets a list of all functional blocks in the simulation.
-        /// </summary>
-        public IReadOnlyList<FunctionalBlock> FunctionalBlocks => functionalBlocks;
+        ///// <summary>
+        ///// Gets a list of all functional blocks in the simulation.
+        ///// </summary>
+        //public IReadOnlyList<FunctionalBlock> FunctionalBlocks => functionalBlocks;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/Simulation.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome/Simulation.cs [hunk 8]
```diff
@@ -48,50 +51,57 @@ public sealed class Simulation : IDisposable
...
         public IGameService Service { get; }
 
-        /// <summary>
-        /// Gets a list of all entities in the simulation.
-        /// </summary>
-        public IReadOnlyList<Entity> Entities => entities;
+        ///// <summary>
+        ///// Gets a list of all entities in the simulation.
+        ///// </summary>
+        //public IReadOnlyList<Entity> Entities => entities;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/Simulation.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Network.Tests/OctoNetworkStreamTest.cs [hunk 9]
```diff
@@ -11,8 +11,11 @@ namespace OctoAwesome.Network.Tests
     public class OctoNetworkStreamTest
     {
         private OctoNetworkStream testStream;
-        private Random rand;
+        private Random rand;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Network.Tests/OctoNetworkStreamTest.cs]:
```suggestion
        private Random rand;
```

File: OctoAwesome/OctoAwesome.Network.Tests/SimulationManagerTests.cs [hunk 10]
```diff
@@ -15,8 +15,11 @@ namespace OctoAwesome.Network.Tests
     [TestOf(typeof(SimulationManager))]
     public class SimulationManagerTests
     {
-        private readonly SimulationManager simulationManager;
+        private readonly SimulationManager simulationManager;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Network.Tests/SimulationManagerTests.cs]:
```suggestion
        private readonly SimulationManager simulationManager;
```

File: OctoAwesome/OctoAwesome.Tests/CollisionTests.cs [hunk 11]
```diff
@@ -14,8 +14,11 @@ public class CollisionTests
     {
         private float gap = 0.0000f;
         private Player Player;
-        private List<Index3> blocks = new List<Index3>();
+        private List<Index3> blocks = new List<Index3>();
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Tests/CollisionTests.cs]:
```suggestion
        private List<Index3> blocks = new List<Index3>();
```

File: OctoAwesome/OctoAwesome.Tests/CollisionTests.cs [hunk 12]
```diff
@@ -14,8 +14,11 @@ public class CollisionTests
     {
         private float gap = 0.0000f;
         private Player Player;
-        private List<Index3> blocks = new List<Index3>();
+        private List<Index3> blocks = new List<Index3>();        
 
+        /// <summary>
+        /// Initializes a new instance of the<see cref="CollisionTests" /> class
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Tests/CollisionTests.cs]:
```suggestion
        /// Initializes a new instance of the <see cref="CollisionTests" /> class.
```

File: OctoAwesome/OctoAwesome.Tests/LocalChunkCacheTest.cs [hunk 13]
```diff
@@ -12,8 +12,11 @@ namespace OctoAwesome.Tests
 
     public class LocalChunkCacheTest
     {
-        TestGlobalCache globalCache;
+        TestGlobalCache globalCache;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Tests/LocalChunkCacheTest.cs]:
```suggestion
        TestGlobalCache globalCache;
```

File: OctoAwesome/OctoAwesome.Tests/LocalChunkCacheTest.cs [hunk 14]
```diff
@@ -12,8 +12,11 @@ namespace OctoAwesome.Tests
 
     public class LocalChunkCacheTest
     {
-        TestGlobalCache globalCache;
+        TestGlobalCache globalCache;        
 
+        /// <summary>
+        /// Initializes a new instance of the<see cref="LocalChunkCacheTest" /> class
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Tests/LocalChunkCacheTest.cs]:
```suggestion
        /// Initializes a new instance of the <see cref="LocalChunkCacheTest" /> class.
```

File: OctoAwesome/OctoAwesome.Tests/TestGlobalCache.cs [hunk 15]
```diff
@@ -28,8 +28,11 @@ public int DirtyChunkColumn
             }
         }
 
-        public IPlanet Planet => throw new NotImplementedException();
+        public IPlanet Planet => throw new NotImplementedException();
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Tests/TestGlobalCache.cs]:
```suggestion
        public IPlanet Planet => throw new NotImplementedException();
```

File: OctoAwesome/OctoAwesome.Tests/TestGlobalCache.cs [hunk 16]
```diff
@@ -28,8 +28,11 @@ public int DirtyChunkColumn
             }
         }
 
-        public IPlanet Planet => throw new NotImplementedException();
+        public IPlanet Planet => throw new NotImplementedException();        
 
+        /// <summary>
+        /// Initializes a new instance of the<see cref="TestGlobalCache" /> class
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Tests/TestGlobalCache.cs]:
```suggestion
        /// Initializes a new instance of the <see cref="TestGlobalCache" /> class.
```

File: OctoAwesome/OctoAwesome/ComponentContainer.cs [hunk 17]
```diff
@@ -147,8 +167,8 @@ protected virtual void OnAddComponent(TComponent component)
         /// </exception>
         protected virtual void ValidateAddComponent(TComponent component)
         {
-            if (Simulation is not null)
-                throw new NotSupportedException("Can't add components during simulation");
+            //if (Simulation is not null)
+            //    throw new NotSupportedException("Can't add components during simulation");
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/ComponentContainer.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome/ComponentContainer.cs [hunk 18]
```diff
@@ -160,8 +180,8 @@ protected virtual void ValidateAddComponent(TComponent component)
         /// </exception>
         protected virtual void ValidateRemoveComponent(TComponent component)
         {
-            if (Simulation is not null)
-                throw new NotSupportedException("Can't remove components during simulation");
+            //if (Simulation is not null)
+            //    throw new NotSupportedException("Can't remove components during simulation");
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/ComponentContainer.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/Definitions/Items/FurnaceItem.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+            simulationRelay.OnNext(notification);
+            return 0;
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            simulationSource.Dispose();
+            simulationRelay?.Dispose();
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/Definitions/Items/FurnaceItem.cs]:
```suggestion
            simulationRelay.Dispose();
```

File: OctoAwesome/OctoAwesome.Basics/Extension.cs [hunk 20]
```diff
@@ -98,34 +107,119 @@ public void Register(IExtensionLoader extensionLoader, ITypeContainer typeContai
...
-                    c.transferUiComponent.Closed += c.TransferUiComponentClosed;
-                    c.Components.AddComponent(c.transferUiComponent, true);
-                }
+                //if (!c.ContainsComponent<TransferUIComponent>())
+                //{
+                //    c.transferUiComponent = new TransferUIComponent(inventoryComponent);
+                //    c.transferUiComponent.Closed += c.TransferUiComponentClosed;
+                //    //TODO: Fix this
+                //    //c.Components.AddComponent(c.transferUiComponent, true);
+                //}
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/Extension.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/Extension.cs [hunk 21]
```diff
@@ -98,34 +107,119 @@ public void Register(IExtensionLoader extensionLoader, ITypeContainer typeContai
...
+                }
+
+                //if (!c.ContainsComponent<TransferUIComponent>())
+                //{
+                //    c.transferUiComponent = new TransferUIComponent(inventoryComponent);
+                //    c.transferUiComponent.Closed += c.TransferUiComponentClosed;
+                //    //TODO: Fix this
+                //    //c.Components.AddComponent(c.transferUiComponent, true);
+                //}
+
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/Extension.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/FunctionBlocks/Chest.cs [hunk 22]
```diff
@@ -13,7 +16,9 @@ namespace OctoAwesome.Basics.FunctionBlocks
     public class Chest : FunctionalBlock
     {
         internal AnimationComponent animationComponent;
-        internal TransferUIComponent transferUiComponent;
+        private IDisposable changedSub;
+
+        //internal TransferUIComponent transferUiComponent;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/FunctionBlocks/Chest.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/UI/Components/FurnaceUIComponent.cs [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+        //target.AddSlot(slot, 0);
+        //source.AddSlot(firstSlot);
+        //}
+        //else
+        //{
+        //    var inputOrOutput = source.Inventory.IndexOf(slot);
+        //    source.RemoveSlot(slot);
+        //    target.AddSlot(slot);
+        //    source.AddSlot(new InventorySlot(), inputOrOutput);
+        //}
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/UI/Components/FurnaceUIComponent.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/UI/Components/TransferUIComponent.cs [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+        if (toAddAndRemove == 0)
+            return;
+        var item = slot.Item;
+        var amount = source.Remove(slot, toAddAndRemove);
+
+        var addedAddedAmount = target.Add(item, toAddAndRemove);
+        Debug.Assert(amount == addedAddedAmount, "The added value and removed value of the inventories is unequal, threading?");
+        //sourceControl.Rebuild(source.Inventory);
+        //targetControl.Rebuild(target.Inventory);
+
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/UI/Components/TransferUIComponent.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/UI/Controls/FurnaceControl.cs [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+        //    var label = new Label(ScreenManager) { Text = inventorySlot.Amount.ToString(), HorizontalAlignment = HorizontalAlignment.Right, VerticalTextAlignment = VerticalAlignment.Bottom, Background = new BorderBrush(Color.White) };
+
+        //    var grid = new Grid(ScreenManager);
+        //    grid.Columns.Add(new ColumnDefinition() { ResizeMode = ResizeMode.Parts, Width = 1 });
+        //    grid.Rows.Add(new RowDefinition() { ResizeMode = ResizeMode.Fixed, Height = 50 });
+        //    grid.AddControl(image, 0, 0);
+        //    grid.AddControl(label, 0, 0);
+
+        //    outputPanel.Controls.Add(grid);
+        //}
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/UI/Controls/FurnaceControl.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/UI/Screens/FurnaceScreen.cs [hunk 26]
```diff
@@ -0,0 +1,217 @@
...
+            grid.AddControl(furnace, 0, 2);
+
+            var infoPanel = new StackPanel(manager)
+            {
+                HorizontalAlignment = HorizontalAlignment.Stretch,
+                VerticalAlignment = VerticalAlignment.Stretch,
+                Background = NineTileBrush.FromSingleTexture(panelBackground, 30, 30),
+                Padding = Border.All(20),
+                Margin = Border.All(10, 0, 0, 0),
+
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/UI/Screens/FurnaceScreen.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/UI/Screens/FurnaceScreen.cs [hunk 27]
```diff
@@ -0,0 +1,217 @@
...
+            {
+                e.Handled = true;
+                var source = slot.GetParentInventory();
+                furnaceUIComponent.Transfer(source, target, slot);
+
+                //if (transferDirection == TransferDirection.AToB)
+                //else if (transferDirection == TransferDirection.BToA)
+                //    furnaceUIComponent.Transfer(furnaceUIComponent.InputInventory, furnaceUIComponent.InventoryA, slot);
+                //else
+                //    Debug.Fail($"{nameof(transferDirection)} has to be {nameof(TransferDirection.AToB)} or {nameof(TransferDirection.BToA)}");
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/UI/Screens/FurnaceScreen.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/UI/Screens/TransferScreen.cs [hunk 28]
```diff
@@ -105,54 +116,111 @@ public TransferScreen(BaseScreenComponent manager, AssetComponent assetComponent
...
+        private void OnInventoryDrop(DragEventArgs e, InventoryComponent target)
+        {
+            if (transferComponent is not null && e.Content is InventorySlot slot)
             {
+                var source = slot.GetParentInventory();
                 e.Handled = true;
-                MoveSlot(slot, sourceControl, source, targetControl, target);
+                transferComponent.Transfer(source, target, slot);
+                //if (source.RemoveSlot(slot))
+                //    target.AddSlot(slot);
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/UI/Screens/TransferScreen.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Basics/UI/Screens/TransferScreen.cs [hunk 29]
```diff
@@ -105,54 +116,111 @@ public TransferScreen(BaseScreenComponent manager, AssetComponent assetComponent
...
+            var item = slot.Item;
+            var amount = source.Remove(slot, toAddAndRemove);
 
-            target.AddSlot(slot);
-            sourceControl.Rebuild(source.Inventory);
-            targetControl.Rebuild(target.Inventory);
+            var addedAddedAmount = target.Add(item, toAddAndRemove);
+            Debug.Assert(amount == addedAddedAmount, "The added value and removed value of the inventories is unequal, threading?");
+            //sourceControl.Rebuild(source.Inventory);
+            //targetControl.Rebuild(target.Inventory);
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/UI/Screens/TransferScreen.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Client.UI/Components/UIComponent.cs [hunk 30]
```diff
@@ -0,0 +1,323 @@
...
+
+        /// <summary>
+        /// Displays the ui
+        /// </summary>
+        public abstract void Update(GameTime gameTime);
+
+        /// <inheritdoc/>
+        public void Add(ComponentContainer value)
+        {
+
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client.UI/Components/UIComponent.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Client.UI/Screens/BaseScreen.cs [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+        private readonly Button backButton;
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BaseScreen"/> class.
+        /// </summary>
+        public BaseScreen(BaseScreenComponent manager, AssetComponent assets) : base(manager)
+        {
+            this.assets = assets;
+            backButton = new TextButton(Manager, "Back"
+                //UI.Languages.OctoClient.Back
+                )
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client.UI/Screens/BaseScreen.cs]:
```suggestion
            backButton = new TextButton(Manager, "Back")
```

File: OctoAwesome/OctoAwesome.Client/Components/ScreenComponent.cs [hunk 32]
```diff
@@ -1,25 +1,57 @@
-﻿using engenious.UI;
+﻿using engenious;
+using engenious.UI;
+using engenious.UI.Controls;
+
 using OctoAwesome.Client.Screens;
 using System;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Components/ScreenComponent.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Client/Components/ScreenComponent.cs [hunk 33]
```diff
@@ -1,25 +1,57 @@
...
 using OctoAwesome.Client.UI.Components;
+using OctoAwesome.UI.Components;
+using OctoAwesome.Database;
+using OctoAwesome.Extension;
+using OctoAwesome.Notifications;
+using OctoAwesome.Runtime;
+using OctoAwesome.Rx;
+using OctoAwesome.UI.Screens;
+
+using System.Collections.Generic;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Components/ScreenComponent.cs:16]:
```suggestion
using System;
using System.Collections.Generic;
```

File: OctoAwesome/OctoAwesome/Definitions/IFoodMaterialDefinition.cs [hunk 34]
```diff
@@ -0,0 +1,24 @@
...
+        /// <summary>
+        /// Determines if this food can be eaten, independent of the flavor
+        /// </summary>
+        bool Edible { get; }
+
+        //TODO move to item
+        ///// <summary>
+        ///// Can this be eaten without beeing poisened or not
+        ///// </summary>
+        //bool Poisenous { get; }
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/Definitions/IFoodMaterialDefinition.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Client/Control/ExtensionsOptionControl.cs [hunk 35]
```diff
@@ -118,7 +119,7 @@ public ExtensionsOptionControl(BaseScreenComponent manager, IExtensionLoader ext
             applyButton.LeftMouseClick += (s, e) =>
             {
                 //TODO: Apply
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Client/Control/ExtensionsOptionControl.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome/EntityComponents/InventoryComponent.cs [hunk 36]
```diff
@@ -115,85 +273,456 @@ public override void Serialize(BinaryWriter writer)
...
+        //        {
+        //            Amount = inventorySlot.Amount,
+        //        };
+        //        inventory.Add(slot);
+        //    }
+        //    else
+        //    {
+        //        slot.Amount += inventorySlot.Amount;
+        //    }
+        //}
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/EntityComponents/InventoryComponent.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome/Crafting/RecipeService.cs [hunk 37]
```diff
@@ -0,0 +1,256 @@
...
+    //    }
+
+    //    return null;
+    //}
+
+    //TODO Is this needed, when match is finally implemented
+    //public IReadOnlyCollection<Recipe> Search(string query)
+    //{
+    //    throw new NotImplementedException();
+    //}
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/Crafting/RecipeService.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome/EntityComponents/InventoryComponent.cs [hunk 38]
```diff
@@ -115,85 +273,456 @@ public override void Serialize(BinaryWriter writer)
...
+        //{
+        //    int alreadyRemoved = 0;
+        //    var slots = inventory.Where(x => x.Item == item);
+
+        //    foreach (var slot in slots)
+        //    {
+        //        Remove(item, slot);
+
+        //    }
+        //}
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/EntityComponents/InventoryComponent.cs]:
```suggestion
```

File: OctoAwesome/OctoAwesome/EntityComponents/InventoryComponent.cs [hunk 39]
```diff
@@ -115,85 +273,456 @@ public override void Serialize(BinaryWriter writer)
...
+
+        ✓ 13. Get Inventory Fullness (Public prop getter for current and max)
+        ✓ 14. How much of X can be added to Inventory
+
+
+        ✓ TODO Rework serialize / deserialize to support the new inventory structure
+         
+        ✓ TODO Add all slots on initialize for fixed size inventory
+        ✓ TODO Where and How to Initialize Limits
+
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/EntityComponents/InventoryComponent.cs]:
```suggestion
```

File: OctoAwesome/Design Documents/RecipeOrReceipt.md [hunk 40]
```diff
@@ -0,0 +1,77 @@
...
+* Combined: Mind. 3 Minuten, Zeit 15 Minuten XOR 500 wh für 200g Fleisch
+
+____
+Entscheidung:
+- Required: Zeit XOR Energy
+- Optional: Mindestzeit
+
+Rausgefallen:
+- Required: Mindestzeit
+- Optional: Energy OR Zeit
```
Comment (jvbsl) [inline OctoAwesome/Design Documents/RecipeOrReceipt.md]:
```suggestion
- Optional: Energy OR Zeit 

[RecipeService.cs](https://github.com/OctoAwesome/octoawesome/blob/0a4e469fad1744f5c663b61779e4392d2fb117b1/OctoAwesome/OctoAwesome/Crafting/RecipeService.cs)
```csharp

//TODO Implement on a later point when needed or all discrepancies are solved
public Recipe? Match(IReadOnlyCollection<Recipe> recipes, IReadOnlyCollection<RecipeItem> inputs, IReadOnlyCollection<RecipeItem> outputs, string category = "", string type = "")
{
    foreach (var recipe in recipes.OrderBy(x => x.Inputs.Length))
    {
        if (!string.IsNullOrWhiteSpace(category) && recipe.Category != category)
            continue;
        if (!string.IsNullOrWhiteSpace(type) && recipe.Type != type)
            continue;

        int counter = 0;
        foreach (var inputItem in recipe.Inputs)
        {
            foreach (var input in inputs)
            {
                if (inputItem.Count <= input.Count
                    && (string.IsNullOrWhiteSpace(inputItem.ItemName) || inputItem.ItemName == input.ItemName)
                    && (string.IsNullOrWhiteSpace(inputItem.MaterialName) || inputItem.MaterialName == input.MaterialName))
                {
                    counter++;
                    break;
                }
            }
        }
        if (counter >= recipe.Inputs.Length)
            return recipe;
    }

    return null;
}

// TODO Is this needed, when match is finally implemented
public IReadOnlyCollection<Recipe> Search(string query)
{
    throw new NotImplementedException();
}
```
```

File: OctoAwesome/OctoAwesome.Basics/EntityComponents/UIComponents/TransferUIComponent.cs [hunk 41]
```diff
@@ -1,61 +1,51 @@
...
-                return;
-
-            transferScreen.Rebuild(chestInventory, playerInventory);
-
-            if (ScreenComponent.ActiveScreen != transferScreen)
-                ScreenComponent.NavigateToScreen(transferScreen);
-        }
-    }
-}
+﻿//using engenious.UI;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Basics/EntityComponents/UIComponents/TransferUIComponent.cs]: Delete file instead

File: OctoAwesome/OctoAwesome/FunctionalBlock.cs [hunk 42]
```diff
@@ -18,11 +31,5 @@ public void Interact(GameTime gameTime, Entity entity)
             OnInteract(gameTime, entity);
         }
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/FunctionalBlock.cs:33]: Interact does the same as ComponentContainer<IFunctionalBlockComponent>.Interact which is hidden by this method(as the method in the base class is not virtual).
-> Delete Interact method from FunctionalBlock

File: OctoAwesome/OctoAwesome/ComponentContainerList.cs [hunk 43]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/ComponentContainerList.cs]: Class is not in use at all -> Remove file

**Review by jvbsl — CHANGES_REQUESTED** [review]
Some things need to be refactored slightly

File: OctoAwesome/OctoAwesome/FunctionalBlock.cs (hunk 42 above)
Comment (susch19) [reply OctoAwesome/OctoAwesome/FunctionalBlock.cs:33]: FunctionalBlock and Entity will be merged pretty soon in the future, so this will resolve itself

**Review by jvbsl — APPROVED** [review]

**Pull request merged**